### PR TITLE
KAFKA-16637: AsyncKafkaConsumer removes offset fetch responses from cache too aggressively

### PR DESCRIPTION
**Pull request by kirktrue**
Allow the committed offsets fetch to run for as long as needed. This handles the case where a user invokes `Consumer.poll()` with a very small timeout (including zero).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: would `boolean hasPendingOffsetFetch()` be sufficient for the test?

Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: That made sense, so I changed it. Thanks!

**Comment by kirktrue** [comment]
@AndrewJSchofield @cadonna @lianetm @philipnee—please review this PR. It's an alternative take on #16241 that seems simpler and not fraught with peril.

Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1688]: The approach looks good but `getResult` can throw exceptions. There are three cases to consider I think:

* TimeoutException - This is the case this PR is aiming to fix, so I think this is fine.
* InterruptedException - I think you can also leave the pending event in place.
* Everything else - Probably ought to clear the pendingOffsetFetchEvent because the event is doomed.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the PR. I like the new approach and the PR looks pretty good. I have left one comment to do with exception handling.

**Review by cadonna — COMMENTED** [review]
Thanks for the PR, @kirktrue !

I like the approach.

Once the @AndrewJSchofield's comments are addressed, I guess we are good to go.

WDYT, @lianetm ?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (lianetm) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1688]: Agree with this point. We could probably catch whatever the getResult throws, if not timeout or interrupt clear pendingOffsetFetchEvent, and then throw whatever we got.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: uhm actually here we expect that a Fetch event was never added (because it should reuse the previous one). We could have `verify(applicationEventHandler, never()).add(any(FetchCommittedOffsetsEvent.class));` here to make the intention clear, and `clearInvocations(applicationEventHandler);` before the poll (I know the logic it's right but this line gives the wrong impression and makes me doubt our logic is right, and re-read the whole test :) )

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java
Comment (lianetm) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: just fyi, here I agree it makes sense to have it like this without clearing invocations on the applicationEventHandler, as we see how new events are generated on subsequent polls, nice.

**Comment by lianetm** [comment]
Thanks for the PR @kirktrue, I love the simple the approach and the consistency with the legacy logic, thanks for your patience here :)

Same as @cadonna , LGTM once the error handling on getResult that @AndrewJSchofield pointed out is addressed.

Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1688]: I want to clarify the desired error handling logic:

- `TimeoutException`: log message, keep pending event, return `false`
- `InterruptException`: no log, keep pending event, throws error
- Everything else - no log, clears pending event, throws `KafkaException`-wrapped error

Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1688]: Is there any _harm_ in having `InterruptException` clear out the pending event? Why else would an interrupt exception be thrown in normal operations?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java
Comment (kirktrue) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: I hate this, BTW. Suggestions welcomed.

Comment (AndrewJSchofield) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: I suspect it's just one of those things.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java
Comment (kirktrue) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: IntelliJ's editor highlights this line and warns: `Result of 'maybeWrapAsKafkaException()' not thrown`. Suggestions on how to rewrite this welcomed.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1688]: There's no _harm_ in having `InterruptException` clear out the pending event.

**Comment by AndrewJSchofield** [comment]
I'm happy with the approach now. Thanks @kirktrue.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java
Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: honestly no idea ;), but well, the whole complexity I see in this test is about mocking this static consumerUtils. Another alternative (simpler) would be to just complete the future of the fetch event with the exception we want to test right? it also removes the thread interrupt code you did not like above (that agree with @AndrewJSchofield , it is what it is). Happy to keep it like this, but if you want to try the alternative I imagine it would be along the lines of :
```
// check timeout
            event1.future().completeExceptionally(new TimeoutException(""));
            assertDoesNotThrow(() -> consumer.poll(Duration.ofMillis(timeoutMs)));
// check interrupt
            event1.future().completeExceptionally(new InterruptException(""));
            assertThrows(reportedFailure.getClass(), () -> consumer.poll(Duration.ofMillis(timeoutMs)));
```
(no  `consumerUtilsMockedStatic` needed at all)

**Comment by lianetm** [comment]
Thanks for the updates @kirktrue , happy with the approach too. Left an answer to your concerns on the tests, can be totally done separately if you prefer.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java
Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: agree it is what it is if needed...(but we can maybe avoid it with my suggestion below?)

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: @lianetm—thanks for the suggestion.

I was using `Future.completeExceptionally()` at first, but I couldn't get it to work because when `Future.get()` is later invoked by `ConsumerUtils.getResult()`, the exception is always wrapped in an `ExecutionException`. Even the `InterruptedException` and `TimeoutException` were wrapped in the `ExecutionException`, so I wasn't able to test the other `catch` blocks in `ConsumerUtils.getResult()`.

I also attempted to "capture" the `Future` that was enqueued into the `applicationEventQueue` and force its `get()` method to throw my desired exception type, but I wasn't able to get that to work, either 🤷‍♂️

Comment (lianetm) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: What about this (passes locally for me):
```
        @Test
    public void testOffsetFetchInterrupt() {
        consumer = newConsumer();
        long timeoutMs = 0;
        doReturn(Fetch.empty()).when(fetchCollector).collectFetch(any(FetchBuffer.class));
        consumer.assign(Collections.singleton(new TopicPartition("topic1", 0)));

        consumer.poll(Duration.ofMillis(timeoutMs));
        verify(applicationEventHandler).add(any(FetchCommittedOffsetsEvent.class));
        CompletableApplicationEvent<Map<TopicPartition, OffsetAndMetadata>> event = getLastEnqueuedEvent();
        assertTrue(consumer.hasPendingOffsetFetchEvent());

        event.future().completeExceptionally(new InterruptException(""));
        assertThrows(InterruptException.class, () -> consumer.poll(Duration.ofMillis(timeoutMs)));
        assertTrue(consumer.hasPendingOffsetFetchEvent());
    }
```

Similar for Runtime and Timeout but adjusting the expectations:
```
        // Runtime 
        event.future().completeExceptionally(new RuntimeException(""));
        assertThrows(KafkaException.class, () -> consumer.poll(Duration.ofMillis(timeoutMs)));
        assertFalse(consumer.hasPendingOffsetFetchEvent());
        
        // Timeout 
        event.future().completeExceptionally(new TimeoutException(""));
        assertDoesNotThrow(() -> consumer.poll(Duration.ofMillis(timeoutMs)));
        assertTrue(consumer.hasPendingOffsetFetchEvent());
```

All 3 passed for me. (Again, suggesting to simplify getting rid of all the static mocking and warnings, but feel free to leave it for follow-up if you prefer)

Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: Interestingly, that approach is still going through the first catch block in `ConsumerUtils.getResult()`:

```java
try {
    return future.get(timer.remainingMs(), TimeUnit.MILLISECONDS);
} catch (ExecutionException e) {
    throw maybeWrapAsKafkaException(e.getCause());
}
. . .
```

However, because we use the `getCause()`, we end up getting the underlying exception passed to `completeExceptionally()`. The only exception type we end up wrapping is the `NullPointerException` case, which we do want wrapped in a `KafkaException`.

OK, I'll make that change. It's much prettier 🌹

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1688]: _HARM_

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: I've made the tests more clear now. Thanks @lianetm!

**Comment by kirktrue** [comment]
@AndrewJSchofield @cadonna @lianetm @philipnee: this PR is ready to be re-reviewed. Thanks all for your input 😄

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: I switched to using `clearInvocations()` and `never()` for clarity.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: I went ahead and used `clearInvocations()` here too, just for consistency. Is that OK?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumerTest.java]: Avoided the `interrupt()`/`interrupted()` monstrosity by using @lianetm's test approach instead.

**Comment by kirktrue** [comment]
@jlprat—This Jira/PR is a blocker for the KIP-848 Java client work. It sounds like we're really close to merging this within the next day or two. Thanks!

**Review by lianetm — APPROVED** [review]
Thanks for the fix and updates @kirktrue, LGTM.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1725]: Can it get reuse if the partitions of fetch request includes "all" input partitions? It seems `refreshCommittedOffsets` can ignore those partitions.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Does `poll(0)` fill users' log with this error message? if so, should we change the log level or add more explanation in order to avoiding freak users out?

**Review by chia7712 — COMMENTED** [review]
@kirktrue nice patch and approach. two small questions are left. PTAL

**Comment by jlprat** [comment]
Hi @kirktrue. Let's check again when the PR is approved :)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Good point! `ERROR` level seems excessive here. At least `WARN` should be used. However, `WARN` would still fill users' logs. So maybe `DEBUG`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (cadonna) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: nit: Could you please remove the comment? It does not really add any information because it just spells out what the code does.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (cadonna) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: nit:
Why not clearing `pendingOffsetFetchEvent` where you set `shouldClearPendingEvent = true`? 
Would avoid the `if` here and the additional variable `shouldClearPendingEvent`.

**Review by cadonna — APPROVED** [review]
Thanks for all the updates and be open for all the feedback @kirktrue !

LGTM!

However, you should address the concern about excessive logging before we merge.

Thanks to all the other reviewers for their input!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1725]: @chia7712—that's definitely an interesting optimization!

IIUC, the suggestion is to relax the requirement to allow reuse if the partitions for the _current_ request are a subset of (or equal to) the _previous_ request, right? So basically:

```suggestion
        if (!pendingOffsetFetchEvent.partitions().containsAll(partitions))
```

The behavior of the existing `LegacyKafkaConsumer` is to allow reuse only if the partitions for the _current_ request equal those of the _previous_ request **exactly** ([source](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java#L927)). That logic is the basis for the behavior used in the `AsyncKafkaConsumer`. We've been very deliberate to try to match the behavior between the two `Consumer` implementations as closely as possible, unless there's a specific reason not to.

It's a small change, and it does makes sense (to me). My main concern is that it introduces a subtle difference in behavior between the two `Consumer` implementations. Also, the specific case we're trying to solve with this change is when the user has passed in a very low timeout and we're in a tight `poll()` loop, which suggests the partitions wouldn't be changing between those loops (CMIIW).

If I understand correctly, this seems like an optimization, rather than something needed for correctness. If that's the case, can I file a new Jira to implement this when we have a little more time to investigate and test?

Thanks!

Comment (chia7712) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1725]: > can I file a new Jira to implement this when we have a little more time to investigate and test?

sure and thanks for you sharing. I wasn't even aware that "match the behavior" before :smiling_face_with_tear:

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Changed to debug and updated the text of the log to be slightly more helpful.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Removed.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: I went back and forth on this a couple of times. The idea was to have a single place that the variable is assigned and a single place it is cleared, just for easier reasoning of the code for later troubleshooting/debugging/modifying. But it's clearly debatable if it's cleaner that way, so I went ahead and removed the flag.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:1725]: I've filed KAFKA-16966 to track this optimization.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: I'm not sure where that `ERROR`-level debugging crept in. I'd assumed it was a holdover from the `LegacyKafkaConsumer`, but its implementation doesn't log anything in the case of timeouts.

I think it's helpful to keep it there in `DEBUG` form.

**Comment by kirktrue** [comment]
@AndrewJSchofield @cadonna @chia7712 @lianetm @philipnee: this PR is ready to be re-reviewed. Thanks all for your continued input 😄

**Review by AndrewJSchofield — APPROVED** [review]

**Review by lianetm — APPROVED** [review]
Thanks for the simplification and adjusted log level, that will definitely avoid misleading red flags (been there). LGTM.

**Review by chia7712 — APPROVED** [review]
LGTM

**Comment by kirktrue** [comment]
@jlprat Can this get merged into 3.8.0?

**Comment by chia7712** [comment]
> Can this get merged into 3.8.0?

Sorry that I did not wait for @jlprat feedback before backporting to 3.8 ...

@jlprat I will revert that if it is unsuitable to be in 3.8 :)

**Comment by jlprat** [comment]
Hi @chia7712 we can let it stay in the 3.8 branch

**Pull request merged**
